### PR TITLE
Add `name-property-casing` rule.

### PR DESCRIPTION
**Pull request by armano2**
This PR implements rules proposed in #93

**DONE:**
- [x] Create documtation
- [x] Add tests
- [x] Implement rule

### CODE REVIEW

File: docs/rules/name-property-casing.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  name: 'myComponent'
+}
+```
+
+## :wrench: Options
+
+Default casing is set to `PascalCase`
+
+```
+'vue/name-property-casing': [2, 'camelCase'|'kebab-case'|'PascalCase']
```
Comment (michalsnik) [inline docs/rules/name-property-casing.md]: `camelCase|kebab-case|PascalCase`

File: lib/rules/name-property-casing.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/**
+ * @fileoverview Name property casing for consistency purposes
```
Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Use the same title as in documentation

File: lib/rules/name-property-casing.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  const options = context.options[0]
+  const caseType = ['camelCase', 'kebab-case', 'PascalCase'].indexOf(options) !== -1 ? options : 'PascalCase'
+
+  // ----------------------------------------------------------------------
+  // Public
+  // ----------------------------------------------------------------------
+
+  return utils.executeOnVueComponent(context, (obj) => {
+    const node = obj.properties
+      .filter(item => item.type === 'Property' && item.key.name === 'name' && item.value.type === 'Literal')[0]
```
Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Please wrap this line to make it more readable

File: lib/rules/name-property-casing.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    if (node) {
```
Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Add empty line above `if`

Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Btw. you can also do:
```
if (!node) { return; }
```
To not introduce unnecessary indentation.

File: lib/rules/name-property-casing.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  return str.replace(/(?:^\w|[A-Z]|\b\w)/g, (letter, index) => index === 0 ? letter.toLowerCase() : letter.toUpperCase()).replace(/[\s-]+/g, '')
+}
+
+function pascalCase (str) {
+  str = camelCase(str)
+  return str.length > 0 ? str.charAt(0).toUpperCase() + str.slice(1) : ''
+}
+
+function convertCase (str, caseType) {
+  if (caseType === 'kebab-case') {
```
Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Ideal case for `switch/case`. Or better you can create map with converters:

```
const convertersMap = {
  'kebab-case': kebabCase,
  'camelCase': camelCase,
  'PascalCase': pascalCase
}

function getConverter(name) {
  return convertersMap[name] || pascalCase
}
```

And use it like so:
```
function convertCase (str, caseType) {
  return getConverter(caseType)(str)
}
```

WDYT?

File: lib/rules/name-property-casing.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  return camelCase(str)
+}
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+function create (context) {
+  const options = context.options[0]
+  const caseType = ['camelCase', 'kebab-case', 'PascalCase'].indexOf(options) !== -1 ? options : 'PascalCase'
```
Comment (michalsnik) [inline lib/rules/name-property-casing.js]: Please move the array with possible caseTypes out of the `create` function, it's how it's done in most rules I think.

File: tests/lib/rules/name-property-casing.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        message: 'Property name "foo!bar" is not camelCase.',
+        type: 'Literal',
+        line: 3
+      }]
+    },
+    {
+      filename: 'test.js',
+      code: `
+        new Vue({
+          name: 'foo!bar'
```
Comment (michalsnik) [inline tests/lib/rules/name-property-casing.js]: Please add one more test with `foo_bar`

**Comment by armano2** [comment]
@michalsnik i applied your suggestions & i wrapped few additional line of code.

**Comment by michalsnik** [comment]
I'll release it later :)

**Comment by F3n67u** [comment]
Make name-property-casing fixable is very dangerous, because the fix code only fix the name and leave the reference code same. I think the fix should leave to manual fix, we just show the alert and do nothing else.

**Pull request merged**